proposed/2022/dotnet-nuget-why-proposal.md

```diff
@@ -0,0 +1,140 @@
+
+## Customers
+We are building this for .NET users to be successful with understanding dependency graph for a given package.
+
+## Solution
+NuGet restore operation generates following files under `obj` folder for `PackageReference` style projects.
+- `project.assets.json` file that maintains a project's dependency graph, which is used to make sure that all necessary packages are installed on the computer.
+- `{projectName}.nuget.dgspec.json` file that maintains a project's top-level dependencies along with other metadata.
+- `{projectName}nuget.g.targets` and `{projectName}.nuget.g.props` files define various properties, such as the path to where the packages cache is on the local machine. They also include any MSBuild imports that NuGet packages referenced by the project need.
+
+The contents of `{projectName}.nuget.dgspec.json` and `project.assets.json` will be helpful to understand the dependency graph of a given package but there are few limitations.
+
+-  NuGet restore operation downloads packages during dependency resolution which may not part of the final graph The absence of a package id and version combination in `project.assets.json` file signals that it was downloaded during dependency resolution but not part of the final dependency graph.
+- Packages acquired through `PackageDownload` are not tied to the project in any way beyond acquisition. These packages are not recorded in the final graph that `project.assets.json` maintains.
+
+`dotnet nuget why` command prints the dependency graph of a given package only if it is part of the final graph.
+
+```
+dotnet nuget why [<PROJECT>|<SOLUTION>] package <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>]
+    [--version <VERSION>]
+    [-v|--verbosity <LEVEL>]
+
+dotnet nuget why -h|--help
+```
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution or project is found, an error is thrown.
+
+- PACKAGE_NAME
+
+The package for whom the dependency graph has to be identified. Package name wildcards are not supported.
+
+#### Options
+
+- -f|--framework <FRAMEWORK>
+
+Search package depdnency graph only for a specific target framework.
+
+- -?|-h|--help
+
+Prints out a description of how to use the command.
+
+- --version <VERSION>
+
+Version of the package. See NuGet package versioning.
+
+- -v|--verbosity <VERSION>
+
+Sets the verbosity level of the command. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic]. The default is minimal. For more information, see [LoggerVerbosity](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.framework.loggerverbosity).
+
+#### Examples
+
+- List dependency graph of a package given package id.
+
+```
+dotnet nuget why packageA
```

Comment:
I believe correct syntax for this case would be:
`dotnet nuget why package packageA`

Please fix for other examples too.

Reply:
I had a chat with @chgill-MSFT about this and the feedback is use to `dotnet nuget why` command. What does everyone think?

Reply:
But it doesn't leave us room to grow next time. For example, if we ever want to add `dotnet nuget config why`, `dotnet nuget xxx why` (I made up them for now) then we can't use it anymore, in the future nuget going to add more actions and functionality and it may need clarification in the future.

Reply:
I see where you're coming from @erdembayar . Any additional thoughts @kartheekp-ms and @chgill-MSFT ?

Reply:
We're honestly kind of all over right now, that I'm not sure there's a 100% perfect solution.

I think if we're gonna be a adding a new verb, we'd either be specific to nuget only, or consider whether it makes sense as a first class verb in the SDK altogether. 
So extending something to dotnet nuget config why feels wrong, but we also have dotnet nuget sources already.
Illustrated with examples: 

dotnet nuget why - This is a nuget specific command. `why` applies to whatever seems to be the default for nuget (think dotnet nuget sign).
dotnet nuget package why - This is a nuget specific command and a package specific command? package is a first class noun in dotnet.exe, having it be under nuget explicitly is probably unnecessary. On the other, it's flexible as @erdembayar said.
dotnet package why - this makes sense if package is what we care about the most, but does why apply to other nouns?

Reply:
See https://github.com/NuGet/Home/pull/11875/files#r919109645

Reply:
Based on this conversation: https://github.com/NuGet/Home/pull/11875#discussion_r924901011 , we have decided to move forward with `dotnet nuget why`.